File: importers/c++/src/HSLIC.cpp
/*
------------------------------------------------------------------------------ -
MIT License

Copyright(c) 2016 Yale University

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files(the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions :

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

This is based on the source for the paper

"Multi-scale label-map extraction for texture synthesis"
in the ACM Transactions on Graphics(TOG) - Proceedings of ACM SIGGRAPH 2016,
Volume 35 Issue 4, July 2016
by
Lockerman, Y.D., Sauvage, B., Allegre,
R., Dischler, J.M., Dorsey, J.and Rushmeier, H.

http://graphics.cs.yale.edu/site/publications/multi-scale-label-map-extraction-texture-synthesis

If you find it useful, please consider giving us credit or citing our paper.
------------------------------------------------------------------------------ -

@author : Yitzchak David Lockerman
*/


#include "HSLIC.h"
#include "matio.h"
#include <stdint.h>
#include <algorithm>  
#include <iostream>

template<typename T>
void load_rle_data(T* data_array, size_t number_of_ellements, std::vector<int>& slic_indexes)
{
	auto itter = slic_indexes.begin();

	for (size_t elid = 0; elid < number_of_ellements; elid++)
	{
		T run_lenght = data_array[elid];
		T value = data_array[number_of_ellements + elid];

		itter = std::fill_n(itter, run_lenght, value);
	}

}


HSLIC::HSLIC(const std::string& file) throw(std::exception)
{
	//--------Open the mat file-----------------------------------------------------
	mat_t *matfp;
	matfp = Mat_Open(file.c_str(), MAT_ACC_RDONLY);
	if (NULL == matfp) {
		throw std::exception( (std::string("Unable to open MAT file: ") + file).c_str() );
	}


	//--------Read the shape of the image-------------------------------------------
	{
		matvar_t *image_shape = Mat_VarRead(matfp, "image_shape");
		if (image_shape == NULL)
		{
			Mat_Close(matfp);
			throw std::exception((std::string("File dose not include image_shape. Error while loading file: ") + file).c_str());
		}

		if (image_shape->rank != 2 || image_shape->isComplex ||
			image_shape->dims[0] != 1 || image_shape->dims[1] != 3)
		{
			Mat_VarFree(image_shape);
			Mat_Close(matfp);
			throw std::exception((std::string("image_shape is invalid. Error while loading file: ") + file).c_str());
		}

		if (image_shape->data_type == MAT_T_INT64)
		{
			int64_t* image_shape_data = (int64_t*)image_shape->data;
			rows = image_shape_data[0];
			cols = image_shape_data[1];
		}
		else if (image_shape->data_type == MAT_T_INT32)
		{
			int32_t* image_shape_data = (int32_t*)image_shape->data;
			rows = image_shape_data[0];
			cols = image_shape_data[1];
		}
		else
		{
			Mat_VarFree(image_shape);
			Mat_Close(matfp);
			throw std::exception((std::string("image_shape has unknown type. Error while loading file: ") + file).c_str());
		}

		Mat_VarFree(image_shape);
	}

	//--------Read the atomic superpixels-------------------------------------------
	{
		slic_indexes.resize(rows*cols);

		matvar_t *atomic_SLIC_rle = Mat_VarRead(matfp, "atomic_SLIC_rle");

		if (atomic_SLIC_rle == NULL)
		{
			Mat_Close(matfp);
			throw std::exception((std::string("File does not include atomic_SLIC_rle. Error while loading file: ") + file).c_str());
		}

		if (atomic_SLIC_rle->rank != 2 || atomic_SLIC_rle->isComplex ||
			atomic_SLIC_rle->dims[1] != 2)
		{
			Mat_VarFree(atomic_SLIC_rle);
			Mat_Close(matfp);
			throw std::exception((std::string("atomic_SLIC_rle is invalid. Error while loading file: ") + file).c_str());
		}

		if (atomic_SLIC_rle->data_type == MAT_T_INT64)
		{
			load_rle_data((int64_t*)atomic_SLIC_rle->data, atomic_SLIC_rle->dims[0], slic_indexes);
		}
		else if (atomic_SLIC_rle->data_type == MAT_T_INT32)
		{
			load_rle_data((int32_t*)atomic_SLIC_rle->data, atomic_SLIC_rle->dims[0], slic_indexes);
		}
		else
		{
			Mat_VarFree(atomic_SLIC_rle);
			Mat_Close(matfp);
			throw std::exception((std::string("atomic_SLIC_rle has unknown type. Error while loading file: ") + file).c_str());
		}

		Mat_VarFree(atomic_SLIC_rle);
	}

	//--------Compute the lookup table-----------------------
	int number_of_atomic_superpixels = *std::max_element(slic_indexes.begin(), slic_indexes.end()) + 1;
	atomic_lookup_table.resize(number_of_atomic_superpixels);

	for (int i = 0; i < slic_indexes.size(); i++)
		atomic_lookup_table[slic_indexes[i]].push_back(i);

	//--------Read the HSLIC tree-------------------------------------------
	{
		matvar_t *hslic_tree_var = Mat_VarRead(matfp, "HSLIC");

		if (hslic_tree_var == NULL)
		{
			Mat_Close(matfp);
			throw std::exception((std::string("File does not include HSLIC. Error while loading file: ") + file).c_str());
		}
		try
		{
			load_region_node_from_cell_array(hslic_tree_var, tree_root);
		}
		catch (...)
		{
			Mat_VarFree(hslic_tree_var);
			throw;
		}
		Mat_VarFree(hslic_tree_var);
	}


	Mat_Close(matfp);
}


int HSLIC::get_rows() const
{
	return rows;
}

int HSLIC::get_cols() const
{
	return cols;
}

int HSLIC::get_atomic_superpixels_count() const
{
	return atomic_lookup_table.size();
}


/*
Helper function for superpixels_at_scale. Will return the largest superpixels that are a scall less then "scale."
*/
void recursive_build_superpixel_list_at_scale(const std::vector<HSLICNode>& tree_list, float scale, std::vector<HSLICSuperpixel>& out)
{
	for (HSLICNode node : tree_list)
	{
		if (node.region.scale <= scale || node.children_node.size() == 0)
			out.push_back(node.region);
		else
			recursive_build_superpixel_list_at_scale(node.children_node, scale, out);
	}
}


/*
Returns a vector of HSLICSuperpixel containing all
the superpixels at a given scale
*/
std::vector<HSLICSuperpixel> HSLIC::superpixels_at_scale(float scale) const
{
	std::vector<HSLICSuperpixel> out;
	recursive_build_superpixel_list_at_scale(tree_root, scale, out);
	return out;
}


const std::vector<int>& HSLIC::get_atomic_superpixel_indicator()
{
	return slic_indexes;
}

/*Returns the superpixels at a given scale.
This is an indicator array in row major order */
std::vector<int> HSLIC::superpixels_indicator(const std::vector<HSLICSuperpixel>& superpixels) const
{
	std::vector<int> indicator(rows*cols,-1);
	for (int superpixel_id = 0; superpixel_id < superpixels.size(); superpixel_id++)
	{
		//Go through each atomic_superpixel of superpixel_id
		for (int atomic_superpixel : superpixels[superpixel_id].atomic_superpixels)
		{
			//Mark each index of each superpixel
			for (int index : atomic_lookup_table[atomic_superpixel])
				indicator[index] = superpixel_id;
		}
	}

	return indicator;
}



const std::vector<HSLICNode>& HSLIC::get_superpixel_trees()
{
	return tree_root;
}

File: importers/Matlab/HSLIC.m

%HSLIC Loads the HSLIC Superpixes outputted from "multiscale_extraction"
%   This loads the superpixel tree garnered by our method. The low 
%   level details can be accessed through the properties. The class 
%   also includes functions for high level access. 
%


% -------------------------------------------------------------------------------
% MIT License
% 
% Copyright (c) 2016 Yale University
% 
% Permission is hereby granted, free of charge, to any person obtaining a copy
% of this software and associated documentation files (the "Software"), to deal
% in the Software without restriction, including without limitation the rights
% to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
% copies of the Software, and to permit persons to whom the Software is
% furnished to do so, subject to the following conditions:
% 
% The above copyright notice and this permission notice shall be included in all
% copies or substantial portions of the Software.
% 
% THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
% IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
% FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
% AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
% LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
% OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
% SOFTWARE.
% 
% This is based on the source for the paper 
% 
% "Multi-scale label-map extraction for texture synthesis"
% in the ACM Transactions on Graphics (TOG) - Proceedings of ACM SIGGRAPH 2016,
% Volume 35 Issue 4, July 2016 
% by
% Lockerman, Y.D., Sauvage, B., Allegre, 
% R., Dischler, J.M., Dorsey, J. and Rushmeier, H.
% 
% http://graphics.cs.yale.edu/site/publications/multi-scale-label-map-extraction-texture-synthesis
% 
% If you find it useful, please consider giving us credit or citing our paper.   
% -------------------------------------------------------------------------------
% 
% Created on Mon Jul 08 18:38:55 2013
% 
% @author: Yitzchak David Lockerman

classdef HSLIC

    
    properties
        %The SLIC data at the lowest scale
        %This is an array the size of the image, where each element is the
        %(zerobased) index of the atomic scale superpixel.
        atomic_SLIC 
        
        %A cell array with locations for each base superpixels
        %Each element in the array stores the location of the points within
        %each superpixel. Note that, as a matlab array, the elements are
        %accessed by a 1-based-indexing. Thus superpixel 0 is at element 1
        %and so on.
        atomic_SLIC_table 
        
        %The tree of HSLIC superpixels 
        %This is a cell array of top level superpixels in the hierarchy.
        %Each element contains the following properties:
        % * scale: The scale of the SP
        % * list_of_base_superpixels: An array containing the (zerobased)
        % list of atomic superpixels that make up this higher-level 
        % superpixel. 
        % * children: A cell array of child subpixels.  This child list
        % contains the same structure as HSLIC_tree itself.
        HSLIC_tree 
        
        %The list of all scales
        %This is a sorted list of all scales that contain a superpixel
        %transition.
        scale_list 
    end
    
    methods
        function out = HSLIC(filename)
            %Loads the HSLIC superpixels from a saved file. 
            %A labeling file can be created by using the python program
            %"multiscale_extraction." Please see that program for details.
            %This constructer takes the file path of that output file.
            vars = {'image_shape','atomic_SLIC_rle','HSLIC'};
            vars = load(filename,vars{:});

            %load atomic_SLIC fromt the run length encoding used in the file.
            out.atomic_SLIC = zeros(vars.image_shape(1)*vars.image_shape(2),1,'int32');
            upto = 1;
            for ii = 1:size(vars.atomic_SLIC_rle,1)
                out.atomic_SLIC(upto:upto+vars.atomic_SLIC_rle(ii,1)-1) = vars.atomic_SLIC_rle(ii,2);
                upto=upto+vars.atomic_SLIC_rle(ii,1);
            end
            out.atomic_SLIC = reshape(out.atomic_SLIC,vars.image_shape(2),vars.image_shape(1))';

            %Build a lookup table for the diffrent superpixels 
            lookup_table = zeros(3,vars.image_shape(1),vars.image_shape(2));
            lookup_table(1,:,:) = out.atomic_SLIC;
            [lookup_table(3,:,:),lookup_table(2,:,:)] = ...
                      meshgrid(1:vars.image_shape(2),1:vars.image_shape(1));
            
            lookup_table = reshape(lookup_table,3,[]);
            sorted_tables = sortrows(lookup_table')';
            
            [~,ia,~] = unique(sorted_tables(1,:));
            ia = [ia; size(sorted_tables,2)];
            
            out.atomic_SLIC_table = cell(max(max(out.atomic_SLIC+1)),1);
            for ii = 1:(size(ia,1)-1)
                out.atomic_SLIC_table{ii} = sorted_tables(2:end,ia(ii):ia(ii+1)-1);
            end
            
            out.scale_list = get_all_scales(vars.HSLIC);
            out.HSLIC_tree = vars.HSLIC;
        end
        
        function [ indicator ] = superpixels_indicator(obj,scale )
        %SUPERPIXELS_AT_SCALE Returns a table of the superpixels at a level
        %   This will create a 1-based indicator function for all the
        %   superpixels at a given level. The output will be a 2d integer
        %   array with an superpixel index at each location. Note that the
        %   index given is arbitrary.

            indicator = zeros(size(obj.atomic_SLIC));
            [indicator, ~] = make_indicator(indicator,1,obj.HSLIC_tree,obj,scale);
        end
        
    end
    
end

function scale_list = get_all_scales(HSLIC_node_list)
    %Helper function to create the list of all scales recursively. 
    scale_list = [];

    for subnode = HSLIC_node_list
        scale_list = union(scale_list,subnode{1}.scale);
        scale_list = union(scale_list,get_all_scales(subnode{1}.children));
    end

end

function [indicator,next_id] = make_indicator(indicator,next_id,HSLIC_node_list,obj, scale)
    %Helper function to create the indicator function for a scale. 
    
    for subnode = HSLIC_node_list
        if  subnode{1}.scale <= scale || isempty(subnode{1}.children)
            for spid = subnode{1}.list_of_atomic_superpixels 
                indexes = obj.atomic_SLIC_table{spid+1};
                indicator(indexes(1,:),indexes(2,:)) = next_id;
            end
            next_id = next_id + 1;
        else
            [indicator,next_id] = make_indicator(indicator,next_id,subnode{1}.children,obj,scale);
        end
    end
end
